url29ff2e53f3_nicolas3470_Settlers_of_Catan_AI_tgz-pJ8-robotGameJ8/src/soc/util/SOCGameList.java

```diff
@@ -0,0 +1,551 @@
+            return;
+        if (gl.gameData != null)
+            addGames(gl.gameData.elements(), ourVersion);
+        if (gl.gameInfo != null)
+        {
+            // add games, and/or update canJoin flag of games added via gameData.
+            for (Enumeration gnEnum = gl.gameInfo.keys(); gnEnum.hasMoreElements(); )
+            {
+                String gaName = (String) gnEnum.nextElement();
+                GameInfo gi = (GameInfo) gl.gameInfo.get(gaName);
+                if (gi.opts != null)
+                    addGame(gaName, gi.opts, null, ! gi.canJoin);
+                else
+                    addGame(gaName, null, gi.optsStr, ! gi.canJoin);
+            }
+        }
+    }
+
+    /**
+     * Add several games to this GameList.
+     * Calls {@link #addGame(String, Hashtable, String, boolean)} for each one.
+     * @param gamelist Enumeration of Strings and/or {@link SOCGame}s (mix and match);
+     *          game names may be marked with the prefix
+     *          {@link soc.message.SOCGames#MARKER_THIS_GAME_UNJOINABLE}.
+     *          If gamelist is null, nothing happens.
+     *          If any game already exists (per {@link #isGame(String)}), don't overwrite it;
+     *          at most, clear its canJoin flag.
+     * @param ourVersion Version to check to see if we can join,
+     *          same format as {@link soc.util.Version#versionNumber()}.
+     *          For each SOCGame in gameList, {@link SOCGame#getClientVersionMinRequired()}
+     *          will be called.
+     * @since 1.1.07
+     */
+    public synchronized void addGames(Enumeration gamelist, final int ourVersion)
+    {
+        if (gamelist == null)
+            return;
+
+        while (gamelist.hasMoreElements())
+        {
+            Object ob = gamelist.nextElement();
+            String gaName;
+            Hashtable gaOpts;
+            boolean cannotJoin;
+            if (ob instanceof SOCGame)
+            {
+                gaName = ((SOCGame) ob).getName();
+                gaOpts = ((SOCGame) ob).getGameOptions();
+                cannotJoin = (ourVersion < ((SOCGame) ob).getClientVersionMinRequired());
+            } else {
+                gaName = (String) ob;
+                gaOpts = null;
+                cannotJoin = false;
+            }
+
+            addGame(gaName, gaOpts, null, cannotJoin);
+        }    
+    }
+
+    /**
+     * remove the game from the list
+     *
+     * @param gaName  the name of the game; should not be marked with any prefix.
+     */
+    public synchronized void deleteGame(final String gaName)
+    {
+        D.ebugPrintln("SOCGameList : deleteGame(" + gaName + ")");
+
+        SOCGame game = (SOCGame) gameData.get(gaName);
+
+        if (game != null)
+        {
+            game.destroyGame();
+            gameData.remove(gaName);
+        }
+
+        GameInfo info = (GameInfo) gameInfo.get(gaName);
+        gameInfo.remove(gaName);
+        synchronized (info.mutex)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Object info.mutex synchronized on line 479 should not be re-assigned in the same synchronized block on line 481. After this re-assignment, you lose the reference to the synchronized object.